packages/core/src/DataSchema.ts

```diff
@@ -0,0 +1,173 @@
+                if (sumOffset + bits > maxSumOffset) {
+                    throw new Error(
+                        `Invalid schema, field "${name}" exceeds available storage`
+                    )
+                }
+                const dataIndex = Math.floor(sumOffset / 253)
+                const offset = sumOffset % 253
+                sumOffset += bits
+                return { ...field, dataIndex, offset, bits }
+            } else if (updateBy === 'replace') {
+                if (bits !== 253 - this.config.REPL_NONCE_BITS)
+                    throw new Error(
+                        `Field must be ${
+                            253 - this.config.REPL_NONCE_BITS
+                        } bits`
+                    )
+                if (replOffset + bits > maxReplOffset) {
+                    throw new Error(
+                        `Invalid schema, field "${name}" exceeds available storage`
+                    )
+                }
+
+                const dataIndex =
+                    this.config.SUM_FIELD_COUNT + Math.floor(replOffset / 253)
+                const offset = replOffset % 253
+                replOffset += bits
+                return { ...field, dataIndex, offset, bits }
+            }
+            throw new Error(
+                `Invalid updateBy strategy for field ${name}: "${updateBy}"`
+            )
+        })
+    }
+
+    buildAttestation(change: { name: string; v: bigint }): Attestation {
+        const field: any = this.schema.find((f) => f.name === change.name)
+
+        if (field === undefined) {
+            throw new Error(`${change.name} not found`)
+        }
+        const fieldIndex: number = field.dataIndex
+        const x: bigint = change.v << BigInt(field.offset)
+
+        const maxVal: bigint = (BigInt(1) << BigInt(field.bits)) - BigInt(1)
+
+        if (x > maxVal << BigInt(field.offset)) {
+            throw new Error(`${change.name} exceeds allocated space`)
+        }
+
+        const attestation: Attestation = {
+            fieldIndex,
+            change: x,
+        }
+
+        return attestation
+    }
+
+    buildAttestations(changes: { name: string; v: bigint }[]): Attestation[] {
+        let attestations: Attestation[] = Array(this.schema.length).fill(null)
+        for (const change of changes) {
+            const field: any = this.schema.find((f) => f.name === change.name)
+
+            if (field === undefined) {
+                throw new Error(`${change.name} not found`)
+            }
+
+            const fieldIndex: number = field.dataIndex
+            let v: bigint = change.v << BigInt(field.offset)
+            const maxVal: bigint = (BigInt(1) << BigInt(field.bits)) - BigInt(1)
+
+            if (attestations[fieldIndex] !== null) {
+                v =
+                    change.v +
+                    (field.updateBy === 'sum'
+                        ? (attestations[fieldIndex].change >>
+                              BigInt(field.offset)) &
```

Comment:
Instead of silently taking the lower bits with `&` we should throw an error if an attestation overflows the field bits.

Reply:
This is meant to select the existing `change` value stored in `attestations[]` from previous attestations, so this isn't supposed to be an overflow check. Simply a way to get the current `change` value to be updated.

Reply:
Now that I think about it, we don't need to extract the existing value and overwrite the attestation `change`, so this is an unnecessary step

Reply:
Also, an error is thrown if an attestation overflows the field bits:
```throw new Error(`${change.name} exceeds allocated space`)```

Reply:
Oh okay. This isn't necessary, you can combine the old changes by doing simple addition.

Reply:
Like you can combine the entire field with the shifted change value

Reply:
To summarize, the part of the code that figures out whether there is an attestation overflow or not is the following:
```js
            // Get existing attestation sum value
            let prevVal: bigint = BigInt(0)

            if (field.updateBy === 'sum' && attestations[fieldIndex] !== null)
                prevVal =
                    (attestations[fieldIndex].change >> BigInt(field.offset)) &
                    maxVal

            if (v + prevVal > maxVal << BigInt(field.offset)) {
                throw new Error(`${change.name} exceeds allocated space`)
            }

```
In this, I first get the previous value stored in the attestation using bitwise operations. Then, I add this to the current change value and check if that exceeds the allocated space, which follows the overflow check you suggested.